File: hotel-client/src/app/core/services/notification.service.ts
import { Injectable, Inject, PLATFORM_ID } from '@angular/core';
import { isPlatformBrowser } from '@angular/common';
import { HttpClient } from '@angular/common/http';
import { MatSnackBar } from '@angular/material/snack-bar';
import { BehaviorSubject, Observable } from 'rxjs'; // Fix: Observable import added
import { tap } from 'rxjs/operators';

export interface Notification {
    notificationId: number;
    userId: number;
    message: string;
    type: string;
    isRead: boolean;
    createdAt: string;
}

@Injectable({
    providedIn: 'root'
})
export class NotificationService {
    private apiUrl = 'http://localhost:5254/api/notifications';
    private unreadCountSubject = new BehaviorSubject<number>(0);
    unreadCount$ = this.unreadCountSubject.asObservable();

    private notificationsSubject = new BehaviorSubject<Notification[]>([]);
    notifications$ = this.notificationsSubject.asObservable();

    constructor(
        private http: HttpClient,
        private snackBar: MatSnackBar,
        @Inject(PLATFORM_ID) private platformId: Object
    ) {

        if (isPlatformBrowser(this.platformId)) {
            this.getUnreadCount();
            this.fetchNotifications();
        }
    }

    getNotifications(): Observable<Notification[]> {
        return this.http.get<Notification[]>(this.apiUrl).pipe(
            tap(notes => this.notificationsSubject.next(notes))
        );
    }

    fetchNotifications() {
        this.getNotifications().subscribe();
    }

    getUnreadCount() {
        this.http.get<{ count: number }>(`${this.apiUrl}/unread-count`).subscribe(res => {
            this.unreadCountSubject.next(res.count);
        });
    }

    markAsRead(id: number) {
        return this.http.put(`${this.apiUrl}/${id}/read`, {}).pipe(
            tap(() => {
                this.getUnreadCount();
                this.fetchNotifications();
            })
        );
    }

    markAllAsRead() {
        return this.http.put(`${this.apiUrl}/read-all`, {}).pipe(
            tap(() => {
                this.getUnreadCount();
                this.fetchNotifications();
            })
        );
    }

    deleteNotification(id: number) {
        return this.http.delete(`${this.apiUrl}/${id}`).pipe(
            tap(() => {
                this.getUnreadCount();
                this.fetchNotifications();
            })
        );
    }

    deleteAllNotifications() {
        return this.http.delete(`${this.apiUrl}/all`).pipe(
            tap(() => {
                this.getUnreadCount();
                this.fetchNotifications();
            })
        );
    }

    checkForNewNotifications() {
        this.getNotifications().subscribe(notifications => {
            const unread = notifications.filter(n => !n.isRead);
        });
    }

    public showToast(message: string, type: string = 'Info') {
        let panelClass = 'toast-info';
        if (type === 'Success') panelClass = 'toast-success';
        if (type === 'Warning') panelClass = 'toast-warning';
        if (type === 'Error') panelClass = 'toast-danger';

        this.snackBar.open(message, 'OK', {
            duration: 5000,
            panelClass: [panelClass]
        });
    }
}

File: HotelReservationApi/Services/NotificationService.cs
using HotelReservation.Api.Repositories;
using HotelReservation.Api.Models;

namespace HotelReservation.Api.Services
{
    public class NotificationService
    {
        private readonly INotificationRepository _repo;

        public NotificationService(INotificationRepository repo)
        {
            _repo = repo;
        }

        public Task<List<Notification>> GetUserNotifications(int userId)
            => _repo.GetByUserId(userId);

        public Task<int> GetUnreadCount(int userId)
            => _repo.GetUnreadCount(userId);

        public async Task MarkAsRead(int notificationId, int userId)
        {
            var notification = await _repo.GetById(notificationId);
            if (notification == null || notification.UserId != userId)
                return;

            if (!notification.IsRead)
            {
                notification.IsRead = true;
                await _repo.Save();
            }
        }

        public async Task MarkAllAsRead(int userId)
        {
            var unread = await _repo.GetUnreadByUserId(userId);
            foreach (var n in unread)
            {
                n.IsRead = true;
            }
            if (unread.Any())
            {
                await _repo.Save();
            }
        }

        public async Task CreateNotification(int userId, string message, string type = "Info")
        {
            var note = new Notification
            {
                UserId = userId,
                Message = message,
                Type = type,
                CreatedAt = DateTime.UtcNow,
                IsRead = false
            };
            await _repo.Add(note);
            await _repo.Save();
        }

        public async Task DeleteNotification(int notificationId, int userId)
        {
            var notification = await _repo.GetById(notificationId);
            if (notification != null && notification.UserId == userId)
            {
                await _repo.Delete(notification);
                await _repo.Save();
            }
        }

        public async Task DeleteAllNotifications(int userId)
        {
            await _repo.DeleteAll(userId);
            await _repo.Save();
        }
    }
}
